daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DaffodilUnparseContentHandler.scala

```diff
@@ -0,0 +1,309 @@
+ * is built from information that is passed to the ContentHandler from an XMLReader parser. In
+ * order to receive the uri and prefix information from the XMLReader, the following features
+ * must be set to true on whatever XMLReader is used: http://xml.org/sax/features/namespaces and
+ * http://xml.org/sax/features/namespace-prefixes
+ *
+ * This class, together with the SAXInfosetInputter, uses coroutines to ensure that only one event,
+ * at a time, is passed between the two classes. The following is the general process:
+ *
+ * - an external call is made to parse an XML Documents
+ * - this class receives a StartDocument call, which is the first infosetEvent that is sent to
+ * the SAXInfosetInputter. That event is put on the inputter's queue, this thread is paused, and
+ * that inputter's thread is run
+ * - when the SAXInfosetInputter is done processing an event and is ready for a new event, it
+ * sends the completed event via the coroutine system, and loads it on the contentHandler's
+ * queue, which restarts this thread and pauses that one. In the expected case, the events will
+ * contain no new information, until the unparse is completed.
+ * -  this process continues until the EndDocument method is called. Once that infosetEvent is
+ * sent to the inputter, it signals the end of events coming from the contentHandler. This
+ * ends the unparseProcess and returns the event with the unparseResult and/or any error
+ * information
+ *
+ * @param dp dataprocessor object that will be used to call the parse
+ * @param output outputChannel of choice where the unparsed data is stored
+ */
+class DaffodilUnparseContentHandler(
+  dp: DFDL.DataProcessor,
+  output: DFDL.Output)
+  extends DFDL.DaffodilUnparseContentHandler {
+  private lazy val inputter = new SAXInfosetInputter(this, dp, output)
+  private var unparseResult: DFDL.UnparseResult = _
+  private lazy val infosetEvent: DFDL.SaxInfosetEvent = new DFDL.SaxInfosetEvent
+  private lazy val characterData = new StringBuilder
+  private var prefixMapping: NamespaceBinding = _
+  private lazy val prefixMappingTrackingStack = new MStackOf[NamespaceBinding]
+  private var contentHandlerPrefixMappingUsed = false
+
+  /**
+   * returns null in the case of an DaffodilUnhandledSAXException
+   */
+  def getUnparseResult: DFDL.UnparseResult = unparseResult
+
+  def enableInputterResolutionOfRelativeInfosetBlobURIs(): Unit = inputter.enableResolutionOfRelativeInfosetBlobURIs()
+
+  override def setDocumentLocator(locator: Locator): Unit = {
+    // do nothing
+  }
+
+  override def startDocument(): Unit = {
+    infosetEvent.eventType = One(StartDocument)
+    sendToInputter()
+  }
+
+  override def endDocument(): Unit = {
+    infosetEvent.eventType = One(EndDocument)
+    sendToInputter()
+  }
+
+  override def startPrefixMapping(prefix: String, uri: String): Unit = {
+    contentHandlerPrefixMappingUsed = true
```

Comment:
Does the SAX API say anythign about not mixing use of startPrefixMapping and xmlns in the Attributes? It seems reasonable to only have one or the other, but if they are allowed to be mixed, we might not want to disable the the mapPrefixMapping stuff if we see startPrefixMapping. It's defnitely nice if they can't be mixed, because then we can avoid that stack logic and parsing Attributes for xmlns.

Reply:
So unfortunately it appears to be implementation dependent when namespaces=false and namespace-prefixes=true (it is listed as 'unknown' behavior...which seems to me to be up for interpretation). In the true/true and true/false scenarios, they are explicitly not mixed. In the SAXParserFactory implementation we use for CLI and tests, it doesn't mix them in any scenarios, though. I just thought to play it safe since we are using a shared variable for both. 

Because if they are mixed, I think it would be duplicated information. There would be no reason to have it be separate information.

Reply:
Seems reasonable then to make the assumption. Probably worth adding a comment somewhere (maybe documenting the val contentHandlerPrefixMappingUsed) how this variable is used and that it's making this assumption that these things aren't mixed.